hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/FileIOEngine.java

```diff
@@ -129,20 +129,20 @@ public Cacheable read(BucketEntry be) throws IOException {
     long offset = be.offset();
     int length = be.getLength();
     Preconditions.checkArgument(length >= 0, "Length of read can not be less than 0.");
-    ByteBuffer dstBuffer = ByteBuffer.allocate(length);
+    ByteBuff dstBuff = be.allocator.allocate(length);
     if (length != 0) {
-      accessFile(readAccessor, dstBuffer, offset);
+      accessFile(readAccessor, dstBuff, offset);
       // The buffer created out of the fileChannel is formed by copying the data from the file
       // Hence in this case there is no shared memory that we point to. Even if the BucketCache
       // evicts this buffer from the file the data is already copied and there is no need to
       // ensure that the results are not corrupted before consuming them.
-      if (dstBuffer.limit() != length) {
+      if (dstBuff.limit() != length) {
         throw new IllegalArgumentIOException(
-            "Only " + dstBuffer.limit() + " bytes read, " + length + " expected");
+            "Only " + dstBuff.limit() + " bytes read, " + length + " expected");
       }
     }
-    dstBuffer.rewind();
-    return be.wrapAsCacheable(new ByteBuffer[] { dstBuffer });
+    dstBuff.rewind();
+    return be.wrapAsCacheable(dstBuff);
```

Comment:
Reading this wrapAsCacheable method.  Seems we have only one BucketEntry type now for shared or exclusive types of entries and that uses the refCount associated with it (within BC scope) and this makes it required to have this kind of Composite count mechanism and all now?

Reply:
yep, exclusive HFileBlock will use CompositeRefCnt#innerRefCnt to share with BucketEntry

Reply:
IMHO this is an unwanted sharing and ref count check.